### PR TITLE
feat(react/utils): Add useProxy() hook

### PR DESCRIPTION
**Pull request by sarimarton**
## Related Issues

Follow-up on https://github.com/pmndrs/valtio/discussions/620

## Summary

- [x] Adds useProxy() hook.
- [x] Deprecates valtio/macro and valtio/vacro/vite.
- [ ] Update readme

## Check List

- [ ] `yarn run prettier` for formatting code and docs

### CODE REVIEW

File: src/react-utils/useProxy.ts [hunk 1]
```diff
@@ -20,18 +20,20 @@ import { useLayoutEffect } from 'react'
...
-export const useProxy = <T extends object>(proxy: T) => {
-  const snapshot = useSnapshot(proxy)
+export function useProxy<T extends object>(proxy: T): T {
+  const snapshot = useSnapshot(proxy) as T
 
   let isRendering = true
 
   useLayoutEffect(() => {
+    // This is an intentional hack
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (sarimarton) [inline src/react-utils/useProxy.ts:30]: Does the linter really error on this? Not passing the dep array is legit usage.

Comment (dai-shi) [reply src/react-utils/useProxy.ts:31]: Yes, it does. What it says is useRef instead. It's legit, but not best practice in general?

Comment (sarimarton) [reply src/react-utils/useProxy.ts:31]: Ah yeah, I see the warning, sorry. It's funny. I couldn't help figuring out a linter compliant version:

```tsx
const isRendering = { val: true }

useLayoutEffect(() => {
  isRendering.val = false
})
```

Now that you can call a hack 😜

Comment (dai-shi) [reply src/react-utils/useProxy.ts:31]: Not worth increasing the bundle size. It's a hack anyway for either case, from the react perspective.

**Comment by dai-shi** [comment]
https://codesandbox.io/s/react-typescript-forked-ix38xu

Deprecation message works.
image

**Comment by dai-shi** [comment]
Hmm, I want to reorganize entry points. Will do it in a new PR along with updating readme.

**Review by dai-shi — APPROVED** [review]
Let's merge this. I will probably change `react-utils` to `react/utils` and then export from `utils` too.

File: src/react-utils/useProxy.ts (hunk 1 above)
Comment (sarimarton) [reply src/react-utils/useProxy.ts:31]: Of course, I was just kidding. Thanks @dai-shi for publishing 🥳 and all the great work!

**Pull request merged**